op-plasma/damgr.go

```diff
@@ -2,48 +2,416 @@ package plasma
 
 import (
 	"context"
+	"errors"
+	"fmt"
+	"io"
 
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
 	"github.com/ethereum/go-ethereum/log"
 
+	"github.com/ethereum-optimism/optimism/op-bindings/bindings"
 	"github.com/ethereum-optimism/optimism/op-service/eth"
 )
 
+// ErrPendingChallenge is returned when data is not available but can still be challenged/resolved
+// so derivation should halt temporarily.
+var ErrPendingChallenge = errors.New("not found, pending challenge")
+
+// ErrExpiredChallenge is returned when a challenge was not resolved and derivation should skip this input.
+var ErrExpiredChallenge = errors.New("challenge expired")
+
+// ErrMissingPastWindow is returned when the input data is MIA and cannot be challenged.
+// This is a protocol fatal error.
+var ErrMissingPastWindow = errors.New("data missing past window")
+
+// ErrInvalidChallenge is returned when a challenge event does is decoded but does not
+// relate to the actual chain commitments.
+var ErrInvalidChallenge = errors.New("invalid challenge")
+
+// L1Fetcher is the required interface for syncing the DA challenge contract state.
+type L1Fetcher interface {
+	InfoAndTxsByHash(ctx context.Context, hash common.Hash) (eth.BlockInfo, types.Transactions, error)
+	FetchReceipts(ctx context.Context, blockHash common.Hash) (eth.BlockInfo, types.Receipts, error)
+	L1BlockRefByNumber(context.Context, uint64) (eth.L1BlockRef, error)
+}
+
+// DAStorage interface for calling the DA storage server.
 type DAStorage interface {
-	GetInput(ctx context.Context, key []byte) ([]byte, error)
-	SetInput(ctx context.Context, img []byte) ([]byte, error)
+	GetInput(ctx context.Context, key Keccak256Commitment) ([]byte, error)
+	SetInput(ctx context.Context, img []byte) (Keccak256Commitment, error)
+}
+
+// Config is the relevant subset of rollup config for plasma DA.
+type Config struct {
+	// Required for filtering contract events
+	DAChallengeContractAddress common.Address
+	// The number of l1 blocks after the input is committed during which one can challenge.
+	ChallengeWindow uint64
+	// The number of l1 blocks after a commitment is challenged during which one can resolve.
+	ResolveWindow uint64
 }
 
 type DA struct {
 	log     log.Logger
+	cfg     Config
+	metrics Metricer
+
 	storage DAStorage
-}
+	l1      L1Fetcher
+
+	state *State
 
-type Input struct {
-	Data eth.Data
+	// the latest l1 block we synced challenge contract events from
+	origin eth.BlockID
+	// the latest recorded finalized head as per the challenge contract
+	finalizedHead eth.L1BlockRef
+	// the latest recorded finalized head as per the l1 finalization signal
+	l1FinalizedHead eth.L1BlockRef
+	// flag the reset function we are resetting because of an expired challenge
```

Comment:
What exactly does this integer signify? I see there is a point where we subtract from it but don't actually performa any other action until it is 0.